### PR TITLE
Visioplainte : Liste des occurrences des plages d'ouverture

### PR DESCRIPTION
**Pull request by victormours**
# Contexte

Suite de https://github.com/betagouv/rdv-service-public/pull/4628

Cet endpoint liste les occurrences des plages d'ouvertures, et permet de savoir à quels guichets il faut assigner des agents quand la journée commence.

J'avais l'espoir qu'on puisse ajouter des absences à des guichets pour les fermer, mais actuellement c'est très difficiel d'extraire la soustraction des absences du calculateur de créneau. De toutes façons c'est pas très utile tant qu'on n'a pas implémenté cette user story : https://www.notion.so/rdvs/R-affectation-des-RDV-en-cas-d-indisponibilit-ae096cceb80f4f37af307c62e566ea0e (et du coup j'ai ajouté ça à la fiche notion)

### CODE REVIEW

File: app/controllers/api/visioplainte/plages_ouverture_controller.rb [hunk 1]
```diff
@@ -0,0 +1,34 @@
+class Api::Visioplainte::PlagesOuvertureController < Api::Visioplainte::BaseController
```
Comment (adipasquale) [inline app/controllers/api/visioplainte/plages_ouverture_controller.rb:1]: remarque ouverte : je comprends la logique de ne pas mettre de s final à « PlagesOuverture » mais ça fait quand même bizarre . Les méthodes magiques AR nous fournissent par ex `agent.plage_ouvertures` ce qui est sémantiquement incorrect mais plus programmatiquement attendu. Je n’ai pas d’avis fort dans un sens ou dans l’autre. Je pense que je n’aurais pas été surpris de voir `PlageOuvertures` alors que je suis surpris là. Mais peut-être que pour les usagers de l’API qui n’ont pas l’habitude des conventions rails ça serait l’inverse 🤷 

et question plus ouverte : je me demande si on ne devrait pas être fidèle à notre complexité sous-jacente et appeler cet endpoint `PlagesOuvertureOccurrences` ? J’ai peur que ça induise des malentendus de dire qu’on retourne des plages d’ouvertures alors qu’on retourne des occurences. Si tu vois dans l’API une plage de date par jour, tu te dis peut-être que tu dois créer autant de plages d’ouvertures que de jours

File: app/controllers/api/visioplainte/plages_ouverture_controller.rb [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+  def index
+    if params[:date_debut].blank? || params[:date_fin].blank?
+      errors = ["Vous devez préciser les paramètres date_debut et date_fin"]
```
Comment (adipasquale) [inline app/controllers/api/visioplainte/plages_ouverture_controller.rb:4]: Nitpick : on peut harmoniser la tournure des messages d’erreurs de ce type avec ceux des creneaux par exemple : "Paramètre date_debut et/ou date_fin manquant"

File: app/controllers/api/visioplainte/plages_ouverture_controller.rb [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+      render(json: { errors: errors }, status: :bad_request) and return
+    end
+
+    date_debut = Date.parse(params[:date_debut])
+    date_fin = Date.parse(params[:date_fin])
+
+    plages_ouvertures = plages_ouvertures_scope
```
Comment (adipasquale) [inline app/controllers/api/visioplainte/plages_ouverture_controller.rb:11]: tu as extrait cette méthode pour éviter de déborder en complexité cyclomatique ? je vois que dans l’endpoint des rdvs tu as préféré désactiver ces règles.
Je me dis qu’on peut éventuellement désactiver ces règles pour tous les controllers de l’api visioplainte ?

File: spec/requests/api/visioplainte/plages_ouverture_spec.rb [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+      Agent.joins(:services).where(services: { name: "Gendarmerie Nationale" }).find_by(last_name: "Guichet 2")
+    end
+    let!(:plage_ouverture1) do
+      create(:plage_ouverture, :no_recurrence, agent: agent1, first_day: Date.new(2024, 8, 19), start_time: Tod::TimeOfDay.new(8), end_time: Tod::TimeOfDay.new(12))
+    end
+    let!(:plage_ouverture2) do
+      create(:plage_ouverture, :no_recurrence, agent: agent2, first_day: Date.new(2024, 8, 19), start_time: Tod::TimeOfDay.new(9), end_time: Tod::TimeOfDay.new(12))
+    end
+
+    it "returns plage ouverture for the correct guichet" do
```
Comment (adipasquale) [inline spec/requests/api/visioplainte/plages_ouverture_spec.rb:58]: éventuellement pour bien montrer que ça marche tu peux rajouter un autre test sans filtre sur le guichet_ids dans ce contexte pour vérifier que ça retourne bien les deux occurrences

File: spec/requests/api/visioplainte/plages_ouverture_spec.rb [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+    end
+  end
+
+  context "without required params" do
+    it "returns an error" do
+      get "/api/visioplainte/plages_ouverture", params: { date_debut: "2024-08-19", date_fin: nil }, headers: auth_header
+      expect(response.status).to eq 400
+      expect(response.parsed_body["errors"].first).to eq "Vous devez préciser les paramètres date_debut et date_fin"
+    end
+  end
```
Comment (adipasquale) [inline spec/requests/api/visioplainte/plages_ouverture_spec.rb:72]: éventuellement un dernier test sur l’authentification pour vérifier qu’on n’a pas oublié d’hériter du bon controlleur

File: spec/requests/api/visioplainte/swagger_doc/plage_ouvertures_spec.rb [hunk 6]
```diff
@@ -0,0 +1,36 @@
+require "swagger_helper"
+
+RSpec.describe "Visioplainte API", swagger_doc: "visioplainte/api.json" do # rubocop:disable RSpec/EmptyExampleGroup
```
Comment (adipasquale) [inline spec/requests/api/visioplainte/swagger_doc/plage_ouvertures_spec.rb:3]: par cohérence avec les autres on peut plutôt describe ça comme "Visioplainte Plages d’ouvertures" ou "Visioplainte Occurrence des plages d’ouvertures"

File: spec/requests/api/visioplainte/swagger_doc/plage_ouvertures_spec.rb [hunk 7]
```diff
@@ -0,0 +1,36 @@
...
+  before do
+    travel_to Time.zone.local(2024, 8, 18, 14, 0, 0)
+    load Rails.root.join("db/seeds/visioplainte.rb")
+  end
+
+  path "/api/visioplainte/plages_ouverture" do
+    get "Ces plages d'ouvertures sont les horaires sur lesquelles un guichet est ouvert, et permettent de savoir à quels guichets affecter des agents.\
```
Comment (adipasquale) [inline spec/requests/api/visioplainte/swagger_doc/plage_ouvertures_spec.rb]: `s/lesquelles/lesquels/` ? j’ai un petit doute mais je pense que ça se réfère aux horaires plutôt qu’aux POs

Comment (adipasquale) [reply spec/requests/api/visioplainte/swagger_doc/plage_ouvertures_spec.rb]: Aussi j’aurais tendance à être un peu plus descriptif ici pour expliquer ce qui va être retourné pour des plages récurrentes

File: app/controllers/api/visioplainte/guichets_controller.rb [hunk 8]
```diff
@@ -1,11 +1,13 @@
 class Api::Visioplainte::GuichetsController < Api::Visioplainte::BaseController
-  def index
-    intervenants = Agent.joins(roles: { organisation: :territory }).where(territories: { name: Territory::VISIOPLAINTE_NAME })
+  def self.guichets
```
Comment (adipasquale) [inline app/controllers/api/visioplainte/guichets_controller.rb:2]: ceci n’est pas une demande de changement mais plutôt une ouverture de discussion :) 

On pourrait peut-être extraire ça dans un objet du genre `lib/visioplainte.rb` ou un concern du modèle agent. Ça pourrait devenir un fichier un peu de référence avec des commentaires on décrit un peu la logique de visioplainte, les choix qui ont été faits (le service abandonné en cours d’implém), les raccourcis pris (un guichet = un agent) etc...

File: app/controllers/api/visioplainte/plages_ouverture_controller.rb [hunk 9]
```diff
@@ -0,0 +1,34 @@
...
+
+    if params[:guichet_ids]
+      plages_ouvertures = plages_ouvertures.where(agent_id: params[:guichet_ids])
+    end
+
+    plages_ouverture_occurences = plages_ouvertures.all_occurrences_for(date_debut..date_fin).map do |plage_ouverture, occurrence|
+      {
+        id: plage_ouverture.id,
```
Comment (adipasquale) [inline app/controllers/api/visioplainte/plages_ouverture_controller.rb:19]: c’est un peu perturbant d’avoir plusieurs items avec le même `id` retourné : 

image

on pourrait renommer `id` en `plage_ouverture_id`, mais ça aurait plus de sens dans le contexte où on renommerait l’endpoint en "PlageOuvertureOccurrences" ou qqch comme ça.

**Review by adipasquale — APPROVED** [review]
🚀 testé avec succès en local, j’ai cherché des edge cases avec les dates, ça fonctionne bien. 

Les tests sont très clairs 👍 

Je n’ai aucune remarque bloquante, que des petites discussions ou suggestions. J’ai répété à plusieurs endroits la suggestion d’appeler l’endpoint occurrences mais je suis ok de laisser tel quel aussi.

File: spec/requests/api/visioplainte/swagger_doc/plage_ouvertures_spec.rb [hunk 10]
```diff
@@ -0,0 +1,36 @@
...
+        let(:date_debut) { "2024-08-19" }
+        let(:date_fin) { "2024-08-25" }
+
+        let!(:plage_ouverture_without_recurrence) do
+          create(:plage_ouverture, :no_recurrence, agent: agent, first_day: Date.new(2024, 8, 19), start_time: Tod::TimeOfDay.new(9), end_time: Tod::TimeOfDay.new(12))
+        end
+
+        let(:agent) do
+          Agent.joins(:services).where(services: { name: "Gendarmerie Nationale" }).last
+        end
```
Comment (adipasquale) [inline spec/requests/api/visioplainte/swagger_doc/plage_ouvertures_spec.rb]: il me semble que ces `let!` sont inutiles dans cette spec de doc swagger, j’ai testé de les supprimer en local et je n’ai pas vu la diff

File: app/controllers/api/visioplainte/plages_ouverture_controller.rb (hunk 1 above)
Comment (victormours) [reply app/controllers/api/visioplainte/plages_ouverture_controller.rb:1]: c'est vrai, mais tout ça dépend pas mal des discussions qu'on a eu avec l'équipe visioplainte et du nommage sur lequel on s'est mis d'accord

File: app/controllers/api/visioplainte/plages_ouverture_controller.rb (hunk 3 above)
Comment (victormours) [reply app/controllers/api/visioplainte/plages_ouverture_controller.rb:11]: c'était surtout par réflexe plutôt que pour la complexité cyclomatique

File: spec/requests/api/visioplainte/swagger_doc/plage_ouvertures_spec.rb (hunk 10 above)
Comment (victormours) [reply spec/requests/api/visioplainte/swagger_doc/plage_ouvertures_spec.rb]: bien vu, je l'ai déplacé pour qu'on ai bien des résultats dans la doc
